test/fluree/db/storage/s3_testcontainers_test.clj

```diff
@@ -0,0 +1,177 @@
+(ns fluree.db.storage.s3-testcontainers-test
+  "S3 storage integration tests using testcontainers and LocalStack
+   
+   These tests require Docker to be installed and running. They are tagged with
+   ^:docker meta tag and are excluded from regular CI/CD runs.
+   
+   To run these tests:
+   - All docker tests: clojure -M:docker-tests
+   - Specific test: clojure -M:cljtest -m kaocha.runner --focus fluree.db.storage.s3-testcontainers-test
+   
+   These tests can be included in a weekly CI/CD job using the :docker-tests alias."
+  (:require [clojure.test :refer [deftest is testing use-fixtures]]
+            [cognitect.aws.client.api :as aws]
+            [fluree.db.api :as fluree])
+  (:import [org.testcontainers.containers.localstack LocalStackContainer]
+           [org.testcontainers.containers.localstack LocalStackContainer$Service]))
+
+(def ^:dynamic *localstack-container* nil)
+(def ^:dynamic *s3-endpoint* nil)
+
+(defn start-localstack-container []
+  (let [container (doto (LocalStackContainer. "3.0.2")
+                    (.withServices (into-array [LocalStackContainer$Service/S3])))]
+    (.start container)
+    {:container container
+     :endpoint (.getEndpointOverride container LocalStackContainer$Service/S3)}))
+
+(defn stop-localstack-container [container-info]
+  (when-let [container (:container container-info)]
+    (.stop container)))
+
+(defn localstack-fixture [f]
+  ;; Set up AWS credentials as system properties
+  (System/setProperty "aws.accessKeyId" "test")
+  (System/setProperty "aws.secretAccessKey" "test")
+  (System/setProperty "aws.region" "us-east-1")
+
+  (let [container-info (start-localstack-container)]
+    (try
+      (binding [*localstack-container* (:container container-info)
+                *s3-endpoint* (str (:endpoint container-info))]
+        (f))
+      (finally
+        (stop-localstack-container container-info)))))
+
+(use-fixtures :once localstack-fixture)
+
+(defn create-s3-client []
+  (let [endpoint-url (java.net.URL. *s3-endpoint*)]
+    (aws/client {:api :s3
+                 :endpoint-override {:protocol (keyword (.getProtocol endpoint-url))
+                                     :hostname (.getHost endpoint-url)
+                                     :port (.getPort endpoint-url)}})))
+
+(deftest ^:integration ^:docker s3-testcontainers-basic-test
+  (testing "Basic S3 operations with testcontainers"
+    (let [bucket "fluree-test"
+          client (create-s3-client)]
+
+      ;; Create bucket
+      (aws/invoke client {:op :CreateBucket
+                          :request {:Bucket bucket}})
+
+      ;; Verify bucket exists
```

Comment:
I think these comments are much more useful as nested calls to the `testing` macro. That way failures provide an outline of what went wrong.

```clojure
(testing "S3"
  (testing "a created bucket"
    ...
    (is ... "should exist")
    (testing "connected to Fluree"
      ...
      (is ... "should create a connection")
      (testing "creating a ledger"
        (is "should succeed")
        (testing "and performing basic operations"
          (testing "committing data"
            ...))))))
```